src/ShuttleAutomation.h

```diff
@@ -11,85 +11,293 @@
 #ifndef __AUDACITY_SHUTTLE_AUTOMATION__
 #define __AUDACITY_SHUTTLE_AUTOMATION__
 
+#include <type_traits>
+#include "EffectInterface.h"
 #include "Shuttle.h"
 
+class Effect;
+//! Interface for manipulations of an Effect's settings
+/*!
+ It is meant to be stateless, so all member functions are const
+ */
+class AUDACITY_DLL_API EffectParameterMethods {
+public:
+   virtual ~EffectParameterMethods();
+   virtual void Reset(Effect &effect) const = 0;
+   virtual void Visit(Effect &effect, SettingsVisitor & S) const = 0;
+   virtual void Get(const Effect &effect, CommandParameters & parms) const = 0;
+   virtual bool Set(Effect &effect, const CommandParameters & parms) const = 0;
+};
+
+//! Generates EffectParameterMethods overrides from variadic template arguments.
+/*!
+For each effect parameter, the function...
+   Reset resets it to a default
+   Visit visits it with a SettingsVisitor object
+   Get serializes it to a string
+   Set deserializes it from a string and returns a success flag (if there is
+      failure, parameters might not all be unchanged)
+
+The constructor optionally takes an argument which is a function to be called at
+the end of Reset or Set, and returning a value for Set.
+
+ @tparam EffectType subclass of Effect,
+    expected to define a public static member function FetchParameters,
+    taking EffectType & and EffectSettings &,
+    and returning a pointer to something that holds the parameters
+ @tparam Parameters a pack of non-type template parameters, whose types are
+    specializations of the class template EffectParameter
+
+ TODO in C++20: could use simply `auto ...Parameters` if they are `constexpr`
+*/
+template <typename EffectType, const auto &...Parameters>
+class CapturedParameters : public EffectParameterMethods {
+public:
+   static constexpr auto *fetcher = &EffectType::FetchParameters;
```

Comment:
On this line, using msvc2019 I get:
`error C3535: cannot deduce type for 'auto *' from 'int'`